### PR TITLE
Install packages and update documentation for different call paths for SGX quote generation

### LINKED ISSUE #3843: Document how to use inproc attestation
Open enclave SDK is moving to use out of proc attestation by default with this PR: https://github.com/openenclave/openenclave/pull/3806

We need to document how developers can use in proc attestation now and after the DCAP driver is available in the kernel.

### PR DESCRIPTION
**Pull request by ryanhsu19**
The [original PR](https://github.com/openenclave/openenclave/pull/3806) did not get updated after pushing commits either from the same branch or from another branch, so I have decided to open a new one here with the same description as follows:

Out-of-process quote generation is done by default on Linux (unless both `libsgx-quote-ex` is uninstalled and `SGX_AESM_ADDR` is unset) while in-process quote generation is always done on Windows.

Based on our discussion, we decide not to add another environment variable since it is prone to errors.

This also fixes #3843.

Signed-off-by: Ryan Hsu ryhsu@microsoft.com

### CODE REVIEW

**Comment by ryanhsu19** [comment]
bors try

**Comment by ryanhsu19** [comment]
The force-pushing was intended for rebasing all the four commits, including a merge commit.

**Review by yentsanglee — APPROVED** [review]

**Comment by ryanhsu19** [comment]
bors try

**Comment by ryanhsu19** [comment]
bors try

**Comment by ryanhsu19** [comment]
bors try

**Comment by ryanhsu19** [comment]
bors try

**Comment by ryanhsu19** [comment]
I added a code to verify if `SGX_AESM_ADDR` is set in `samples/attestation/host/host.cpp` and updated both environment variable changes to `samples/attestation/test-samples.cmake` and those to `samples/attestation/Makefile`. CI/CD may also need updates for package installation.

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 1]
```diff
@@ -169,18 +169,18 @@ by working with an AESM service on the local platform.
...
-    * If the DCAP library is configured for in-process quote generation, the DCAP
-    library is chosen, so this default behavior for existing applications is
-    preserved.
-    * Otherwise if the DCAP library is configured for out-of-process quote
-    generation, the quote-ex library will be used, as it supports more evidence
-    formats.
+    * If `SGX_AESM_ADDR` is not set, the DCAP library is used and is configured for in-process quote generation. This preserves the default behavior for the OE SDK before the integration of the quote-ex library.
+    * Otherwise if `SGX_AESM_ADDR` is set, the quote-ex library will be used, as it supports more evidence formats.
+
+On Linux platforms, to be able to use the DCAP library for in-process quote generation, the user running the process needs to be added to the `sgx_prv` group, with the following command:
```
Comment (yentsanglee) [inline docs/DesignDocs/SGX_QuoteEx_Integration.md:182]: "On Linux platforms running kernel 5.11 or later, ..."

File: samples/test-samples.cmake [hunk 2]
```diff
@@ -199,7 +206,8 @@ foreach (i RANGE ${len})
   if (UNIX)
     # The file-encryptor and helloworld are special cases which also
     # work under simulation, so we test that additional scenario here.
-    if (${SAMPLE} MATCHES "(file-encryptor|helloworld)")
+    if (${SAMPLE} MATCHES "(file-encryptor|helloworld)" AND NOT DEFINED
```
Comment (yentsanglee) [inline samples/test-samples.cmake:209]: Why is this change needed?

Comment (ryanhsu19) [reply samples/test-samples.cmake:209]: I probably pasted it accidentally. I'll remove it later.

File: samples/attestation/Makefile [hunk 3]
```diff
@@ -6,6 +6,11 @@
 OE_CRYPTO_LIB := mbedtls
 export OE_CRYPTO_LIB
 
+ifndef SGX_AESM_ADDR
```
Comment (yentsanglee) [inline samples/attestation/Makefile:9]: 'SGX_AESM_ADDR' should only be set at runtime. Not needed at build time. Can this block move to run:?

File: samples/test-samples.cmake [hunk 4]
```diff
@@ -160,8 +160,15 @@ foreach (i RANGE ${len})
     # Build with the CMake package
     message(
       STATUS "Samples test '${SAMPLE}-${CRYPTO_LIB}' with CMake running...")
-    execute_process(COMMAND ${CMAKE_COMMAND} --build ${SAMPLE_BUILD_DIR}
-                            --target run RESULT_VARIABLE TEST_RESULT)
+    if (SAMPLE MATCHES "attestation")
+      execute_process(
+        COMMAND ${CMAKE_COMMAND} -E env SGX_AESM_ADDR=1 ${CMAKE_COMMAND} --build
```
Comment (mingweishih) [inline samples/test-samples.cmake:165]: Now there is an inconsistency between the CMake and Make builds;
the former requires setting the `SGX_AESM_ADDR` when invoking while the latter sets the variable as part of the Makefile.

Given that SGX_AESM_ADDR is an env variable, can we
- Avoid setting the variable by the build script, but updating build instruction the sample's README. Something like:
```
# Use out-of-proc with CMake
SGX_AESM_ADDR=1 cmake ....

# Use in-proc with CMake
SGX_AESM_ADDR=0 cmake ....

# Use out-of-proc with Make
SGX_AESM_ADDR=1 make ....

# Use in-proc with Make
SGX_AESM_ADDR=0 make ....
```

If we want to enforce out-of-proc as OE's default, we should update the `openenclaverc` instead to set the env variable.

File: samples/attestation/Makefile (hunk 3 above)
Comment (ryanhsu19) [reply samples/attestation/Makefile]: As `test-samples.cmake` defines how a sample's tests run, so I moved it there after trying to export it in some targets without success.

File: samples/test-samples.cmake (hunk 4 above)
Comment (ryanhsu19) [reply samples/test-samples.cmake:165]: Based on our discussion we do not enforce it as OE's default call path, but I moved both settings to `test-samples.cmake`.

Comment (mingweishih) [reply samples/test-samples.cmake:165]: Thanks!

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 5]
```diff
@@ -169,18 +169,18 @@ by working with an AESM service on the local platform.
...
-    library is chosen, so this default behavior for existing applications is
-    preserved.
-    * Otherwise if the DCAP library is configured for out-of-process quote
-    generation, the quote-ex library will be used, as it supports more evidence
-    formats.
+    * If `SGX_AESM_ADDR` is not set, the DCAP library is used and is configured for in-process quote generation. This preserves the default behavior for the OE SDK before the integration of the quote-ex library.
+    * Otherwise if `SGX_AESM_ADDR` is set, the quote-ex library will be used, as it supports more evidence formats.
+
+On Linux platforms running kernel 5.11 or later, to be able to use the DCAP library for in-process quote generation, the user running the process needs to be added to the `sgx_prv` group, with the following command:
+    * `sudo usermod -a -G sgx_prv <username>`
```
Comment (yentsanglee) [inline docs/DesignDocs/SGX_QuoteEx_Integration.md:183]: Add reference to SGX DCAP document:

"Form more information, see [SGX DCAP in-proc quote](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/driver/linux/README.kernel.md#apps-with-in-process-quote)."

**Review by bodzhang — CHANGES_REQUESTED** [review]

File: samples/attestation/README.md [hunk 6]
```diff
@@ -275,3 +275,12 @@ To use only the SGX remote functionality:
 * On Linux, use `make runsgxremote` instead of `make run`.
 
 * On Windows, use `ninja runsgxremote` instead of `ninja run`.
+
+## Determining Call Path for SGX Quote Generation
+
+Evidence of format `OE_FORMAT_UUID_SGX_ECDSA` can be generated by either taking the in-process call path or taking the out-of-process call path. However, if you wish to specify the call path it takes to generate a quote, here is what you can do:
+* To perform in-process quote generation, unset the environment variable `SGX_AESM_ADDR` and ensure that the DCAP library is installed.
```
[the remark this answers is not in the document]
Comment (bodzhang) [reply samples/attestation/README.md:282]: For Linux:

If `SGX_AESM_ADDR` is not set, one can run an existing OE app with "out-of-proc" attestation, using $ SGX_AESM_ADDR=1 <app_name>

Adding `SGX_AESM_ADDR=1` in /etc/environment will set SGX_AESM_ADDR for the whole system. I tried to use $ source /etc/environment, but it does not seem to take effect. After ssh session logout/login, the environment variable does take effect. 

If `SGX_AESM_ADDR` is set by default globally, to run an existing OE app with "in-proc" attestation, one can use $ env -r SGX_AESM_ADDR <app_name>

[the remark this answers is not in the document]
Comment (ryanhsu19) [reply samples/attestation/README.md:282]: Thanks! I have also tried out these commands.

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 7]
```diff
@@ -169,18 +169,15 @@ by working with an AESM service on the local platform.
...
-    preserved.
-    * Otherwise if the DCAP library is configured for out-of-process quote
-    generation, the quote-ex library will be used, as it supports more evidence
-    formats.
+An SGX platform can have either the DCAP library or the quote-ex library, or both of them installed.
+
+* If `SGX_AESM_ADDR` is not set, then the DCAP library must be installed. During quote generation, SGX quoting enclaves will be loaded in the application process.
+* If `SGX_AESM_ADDR` is set, then the quote-ex library must be installed. During quote generation, SGX quoting enclaves will be loaded in the SGX AESM service running in kernel mode.
+
+On Linux platforms running kernel 5.11 or later, to be able to use the DCAP library for in-process quote generation, the user running the process needs to be added to the `sgx_prv` group, with the following command:
```
[the remark this answers is not in the document]
Comment (ryanhsu19) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:182]: Thanks, I have extended what was in line 175.

File: docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md [hunk 8]
```diff
@@ -113,3 +113,20 @@ For more information refer to the [Advanced Test Info](AdvancedTestInfo.md) docu
 ## Build and run samples
 
 To build and run the samples, please look [here](/samples/README_Linux.md).
+
```
Comment (bodzhang) [inline docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md:116]: Is this a duplicated file patch?

File: docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md [hunk 9]
```diff
@@ -113,3 +113,20 @@ For more information refer to the [Advanced Test Info](AdvancedTestInfo.md) docu
...
+## Specify the call path for SGX quote generation
```
Comment (bodzhang) [inline docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md:117]: Better to be consistent with the content in samples/BuildSamplesLinux.md

**Review by bodzhang — CHANGES_REQUESTED** [review]

File: docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md (hunk 9 above)
Comment (ryanhsu19) [reply docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md:117]: Good point, thanks! I'll make descriptions in both Getting Started documents match the one in `samples/BuildSamplesLinux.md`.

**Comment by ryanhsu19** [comment]
bors try

**Review by bodzhang — APPROVED** [review]

**Review by yentsanglee — APPROVED** [review]

**Review by qiucwang — APPROVED** [review]
Thanks for the update!
And do we need to update Windows yml as well?

File: docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md [hunk 10]
```diff
@@ -48,7 +48,7 @@ sudo ./sgx_linux_x64_driver.bin
 
 ### 3. Install the Intel and Open Enclave packages and dependencies
 ```bash
-sudo apt -y install clang-8 libssl-dev gdb libsgx-enclave-common libprotobuf10 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave
+sudo apt -y install clang-8 libssl-dev gdb libsgx-enclave-common libsgx-quote-ex libprotobuf10 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave
 ```
```
Comment (radhikaj) [inline docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md:53]: Can you add this text to this doc as well? ## Determine call path for SGX quote generation in attestation sample

In the attestation sample, you can either take the in-process call path or out-of-process call path to generate evidence of format `OE_FORMAT_UUID_SGX_ECDSA`. If you wish to specify the call path it takes to generate a quote, here is what you can do:
* To perform in-process quote generation, unset the environment variable `SGX_AESM_ADDR` and ensure that the DCAP library is installed.
* To perform out-of-process quote generation, set the environment variable `SGX_AESM_ADDR` to any value and ensure that SGX SDK quote-ex Library is installed.

On Linux, if `SGX_AESM_ADDR` is not set, one can run an existing OE app with out-of-process attestation, using `$ SGX_AESM_ADDR=1 <app_name>`.
* If `SGX_AESM_ADDR=1` is added to `/etc/environment` instead, then it will set `SGX_AESM_ADDR` for the whole system. To unset it for the whole system, simply remove the line. These actions require elevated privileges.
* If `SGX_AESM_ADDR` is set by default globally, to run an existing OE app with in-process attestation, one can use `$ env -u SGX_AESM_ADDR <app_name>`.

Please refer to the following document for more information:
* [Attestation: OE SDK Integration with Intel® SGX SDK quote-ex Library for Generation of Evidence in New Formats](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md)

Comment (ryanhsu19) [reply docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md:53]: I added a brief introduction to it and a link to the attestation sample's README file. What do you think?

File: docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md [hunk 11]
```diff
@@ -74,3 +74,9 @@ Open Enclave SDK binary packages can also be [downloaded from GitHub](https://gi
 ### 4. Verify the Open Enclave SDK install
 
 See [Using the Open Enclave SDK](Linux_using_oe_sdk.md) for verifying and using the installed SDK.
+
+### 5. Determine call path for SGX quote generation in attestation sample
+
+In the attestation sample, you can either take the in-process call path or out-of-process call path to generate evidence of format `OE_FORMAT_UUID_SGX_ECDSA`. Please refer to the following README file for more information:
+
+ - [The Attestation Sample](samples/attestation/README.md)
```
Comment (radhikaj) [inline docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md:82]: Can you link to the heading under the sample using README.md#inproc etc?

File: CHANGELOG.md [hunk 12]
```diff
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Add the initial support of cryptographic module loading in SGX enclaves. Refer to the [design document](docs/DesignDocs/CryptoModuleLoadingSupport.md) for more detail.
 - Add the support of getrandom libc API and syscall in enclaves.
+- Add `libsgx-quote-ex`, `sgx-aesm-service` and several SGX AESM plugins to Ansible scripts, which are used to install project requirements.
```
Comment (radhikaj) [inline CHANGELOG.md:16]: Can you also add a link to the attestation out of proc/in proc attestation. Is there a recommendation for what developers  should choose?

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 13]
```diff
@@ -169,18 +169,20 @@ by working with an AESM service on the local platform.
...
+* If `SGX_AESM_ADDR` is not set, then the DCAP library must be installed. During quote generation, SGX quoting enclaves will be loaded in the application process.
+* If `SGX_AESM_ADDR` is set, then the quote-ex library as well as the AESM plugins must be installed. During quote generation, SGX quoting enclaves will be loaded in the SGX AESM service running in kernel mode. The necessary libraries and plugins are listed as follows:
+  * `libsgx-quote-ex`
+  * `sgx-aesm-service`
+  * `libsgx-aesm-ecdsa-plugin`
+  * `libsgx-aesm-pce-plugin`
+  * `libsgx-aesm-quote-ex-plugin`
+
+On Linux platforms running kernel 5.11 or later, to be able to use the DCAP library for in-process quote generation, the user running the process needs to be added to the `sgx_prv` group, with the following command:
+* `sudo usermod -a -G sgx_prv <username>`
```
Comment (radhikaj) [inline docs/DesignDocs/SGX_QuoteEx_Integration.md:183]: At times this group is not created and the developer needs to create the group.

Comment (ryanhsu19) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:183]: Thanks!

File: docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md (hunk 11 above)
Comment (ryanhsu19) [reply docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md]: Sure. I also changed `samples/...` to `/samples/...`.

File: CHANGELOG.md (hunk 12 above)
Comment (ryanhsu19) [reply CHANGELOG.md]: Done. Out-of-process call path will be the default choice. More information is in the design doc and the attestation sample's README.

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 14]
```diff
@@ -169,18 +169,23 @@ by working with an AESM service on the local platform.
...
+
+At times, this group is not created, so the user may need to check if it exists beforehand, with the following command:
```
Comment (yentsanglee) [inline docs/DesignDocs/SGX_QuoteEx_Integration.md:185]: "Check if this group has been created with the following command:
...
"If the group has not been created, create it with the following command:"
...

File: CHANGELOG.md (hunk 12 above)
Comment (radhikaj) [reply CHANGELOG.md]: Thanks Ryan.

**Review by radhikaj — APPROVED** [review]

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 15]
```diff
@@ -169,18 +169,22 @@ by working with an AESM service on the local platform.
...
+  * `libsgx-aesm-ecdsa-plugin`
+  * `libsgx-aesm-pce-plugin`
+  * `libsgx-aesm-quote-ex-plugin`
+
+On Linux platforms running kernel 5.11 or later, to be able to use the DCAP library for in-process quote generation, the user running the process needs to be added to the `sgx_prv` group, with the following command:
+* `sudo usermod -a -G sgx_prv <username>`
+
+Note - if this group has not been created, create it before adding the user to the group.
+
+For more information, see [SGX DCAP in-proc quote](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/driver/linux/README.kernel.md#apps-with-in-process-quote).
```
Comment (bodzhang) [inline docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: The referenced doc does not explain how to create the "sgx_prv" group, if SGX DCAP driver BIN file is not used, for example, in a Linux distro with kernel 5.11.

**Review by bodzhang — CHANGES_REQUESTED** [review]

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (bodzhang) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: @haitaohuang, do you have a suggestion on the instruction to create the "sgx_prv" group, if SGX DCAP driver is not used?

File: docs/DesignDocs/SGX_QuoteEx_Integration.md [hunk 16]
```diff
@@ -169,18 +169,22 @@ by working with an AESM service on the local platform.
...
-    library is chosen, so this default behavior for existing applications is
-    preserved.
-    * Otherwise if the DCAP library is configured for out-of-process quote
-    generation, the quote-ex library will be used, as it supports more evidence
-    formats.
+An SGX platform can have either the DCAP library or the quote-ex library, or both of them installed.
+
+* If `SGX_AESM_ADDR` is not set, then the DCAP library must be installed. During quote generation, SGX quoting enclaves will be loaded in the application process.
+* If `SGX_AESM_ADDR` is set, then the quote-ex library as well as the AESM plugins must be installed. During quote generation, SGX quoting enclaves will be loaded in the SGX AESM service running in kernel mode. The necessary libraries and plugins are listed as follows:
+  * `libsgx-quote-ex`
```
Comment (haitaohuang) [inline docs/DesignDocs/SGX_QuoteEx_Integration.md:176]: AESM service is a daemon (--nodaemon option is also available). Not a kernel mode module.

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (haitaohuang) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: Currently (2.13-2.10) sgx-aesm-service create the group. We are moving that to libsgx-ae-pce in next release. see https://github.com/intel/linux-sgx/issues/678

So in future, regardless in-proc or out-of-proc, you just need install libsgx-ae-pce. Otherwise, to manually create user group: 
```
# groupadd sgx_prv
```

Create  and trigger udev rules:
```
# cat > /etc/udev/rules.d/90-sgx-v40.rules <<EOF
SUBSYSTEM=="misc",KERNEL=="sgx_enclave",MODE="0666",SYMLINK+="sgx/enclave"
SUBSYSTEM=="misc",KERNEL=="sgx_provision",GROUP="sgx_prv",MODE="0660",SYMLINK+="sgx/provision"
EOF
# udevadm trigger

```

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 16 above)
Comment (ryanhsu19) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:176]: I have removed "running in kernel mode."

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (ryanhsu19) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: Is it necessary to create and trigger the userspace dev rules before using in-proc quote generation? It seems that, without doing the second part, simply doing the first part works for me.

**Comment by CyanDevs** [comment]
bors try

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (bodzhang) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: On a Linux system with kernel lower than 5.11 and DCAP driver lower than1.41, sgx_prv is not required for in-proc attestation. @ryanhsu19 , what's the system configuration (SGX driver) when you did the experiment?

File: docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md [hunk 17]
```diff
@@ -113,3 +113,16 @@ For more information refer to the [Advanced Test Info](AdvancedTestInfo.md) docu
...
 
 To build and run the samples, please look [here](/samples/README_Linux.md).
+
+## Determine call path for SGX quote generation in attestation sample
+
+In the attestation sample, you can either take the in-process call path or out-of-process call path to generate evidence of format `OE_FORMAT_UUID_SGX_ECDSA`. If you wish to specify the call path it takes to generate a quote, here is what you can do:
+* To perform in-process quote generation, unset the environment variable `SGX_AESM_ADDR` and ensure that the DCAP library is installed.
+* To perform out-of-process quote generation, set the environment variable `SGX_AESM_ADDR` to any value and ensure that SGX SDK quote-ex Library is installed.
+
+On Linux, if `SGX_AESM_ADDR` is not set, one can run an existing OE app with out-of-process attestation, using `$ SGX_AESM_ADDR=1 <app_name>`.
```
Comment (dthaler) [inline docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md:123]: This is confusing because it only mentions Linux. What about Windows? Put the answer into the .md file.

File: samples/BuildSamplesLinux.md [hunk 18]
```diff
@@ -106,6 +106,19 @@ For example:
 ~/openenclave/share/openenclave/samples/helloworld/build$ make run
 ```
 
+### Determine call path for SGX quote generation in attestation sample
+
+In the attestation sample, you can either take the in-process call path or out-of-process call path to generate evidence of format `OE_FORMAT_UUID_SGX_ECDSA`. If you wish to specify the call path it takes to generate a quote, here is what you can do:
+* To perform in-process quote generation, unset the environment variable `SGX_AESM_ADDR` and ensure that the DCAP library is installed.
+* To perform out-of-process quote generation, set the environment variable `SGX_AESM_ADDR` to any value and ensure that SGX SDK quote-ex Library is installed.
+
+On Linux, if `SGX_AESM_ADDR` is not set, one can run an existing OE app with out-of-process attestation, using `$ SGX_AESM_ADDR=1 <app_name>`.
```
Comment (dthaler) [inline samples/BuildSamplesLinux.md:115]: You don't need to say "On Linux" here because this file is already a Linux specific .md file.

File: samples/attestation/README.md [hunk 19]
```diff
@@ -275,3 +275,16 @@ To use only the SGX remote functionality:
...
+* To perform out-of-process quote generation, set the environment variable `SGX_AESM_ADDR` to any value and ensure that SGX SDK quote-ex Library is installed.
+
+On Linux, if `SGX_AESM_ADDR` is not set, one can run an existing OE app with out-of-process attestation, using `$ SGX_AESM_ADDR=1 <app_name>`.
```
Comment (dthaler) [inline samples/attestation/README.md:285]: This is confusing because it only mentions Linux.  What about Windows?  Put the answer in the .md file.

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (ryanhsu19) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: I opened a new VM with Ubuntu 18.04 LTS and a v5.40 kernel. Then I installed `sgx_linux_x64_driver_1.41.bin` manually.

File: docs/GettingStartedDocs/Contributors/SGX1FLCGettingStarted.md [hunk 20]
```diff
@@ -127,3 +127,16 @@ Follow the instructions in the [Install Info](LinuxInstallInfo.md) document to i
...
 
 To build and run the samples, please look [here](/samples/README.md).
+
+## Determine call path for SGX quote generation in attestation sample
+
+In the attestation sample, you can either take the in-process call path or out-of-process call path to generate evidence of format `OE_FORMAT_UUID_SGX_ECDSA`. If you wish to specify the call path it takes to generate a quote, here is what you can do:
+* To perform in-process quote generation, unset the environment variable `SGX_AESM_ADDR` and ensure that the DCAP library is installed.
+* To perform out-of-process quote generation, set the environment variable `SGX_AESM_ADDR` to any value and ensure that SGX SDK quote-ex Library is installed.
+
+On Linux, if `SGX_AESM_ADDR` is not set, one can run an existing OE app with out-of-process attestation, using `$ SGX_AESM_ADDR=1 <app_name>`.
```
[the remark this answers is not in the document]
Comment (ryanhsu19) [reply docs/GettingStartedDocs/Contributors/SGX1FLCGettingStarted.md]: This documentation is for Linux and I just removed "On Linux." As for Windows, in-process quote generation is recommended, as mentioned in the attestation sample's README file.

File: docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md (hunk 17 above)
Comment (ryanhsu19) [reply docs/GettingStartedDocs/Contributors/SGX1GettingStarted.md]: Same as above.

File: samples/attestation/README.md (hunk 19 above)
Comment (ryanhsu19) [reply samples/attestation/README.md:285]: Added the answer to it. Thanks!

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (haitaohuang) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: if you use the bin installer, it will do those steps for you.

**Comment by ryanhsu19** [comment]
bors try

File: docs/DesignDocs/SGX_QuoteEx_Integration.md (hunk 15 above)
Comment (bodzhang) [reply docs/DesignDocs/SGX_QuoteEx_Integration.md:187]: @haitaohuang, will https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/driver/linux/README.kernel.md#apps-with-in-process-quote be updated to include the information you posted in this thread?

**Review by bodzhang — APPROVED** [review]

**Comment by CyanDevs** [comment]
bors try

**Comment by CyanDevs** [comment]
bors try
